anndata/_core/aligned_mapping.py

```diff
@@ -241,15 +241,14 @@ def to_df(self) -> pd.DataFrame:
         return df
 
     def _validate_value(self, val: V, key: str) -> V:
-        if (
-            hasattr(val, "index")
-            and isinstance(val.index, cabc.Collection)
-            and not (val.index == self.dim_names).all()
-        ):
-            # Could probably also re-order index if it’s contained
-            raise ValueError(
-                f"value.index does not match parent’s axis {self.axes[0]} names"
-            )
+        if hasattr(val, "index") and isinstance(val.index, cabc.Collection):
+            try:
+                pd.testing.assert_index_equal(val.index, self.dim_names)
+            except AssertionError as e:
+                # Could probably also re-order index if it’s contained
+                raise ValueError(
+                    f"value.index does not match parent’s axis {self.axes[0]} names:\n{e}"
+                ) from None
```

Comment:
* Why this change here?
* I would probably go with `val.index.equals(self.dim_names)`. The assertion tests may spend an arbitrary amount of time formatting a nice error message, and aren't built for being performant.

Reply:
Because it would have saved me from jumping into the debugger to see *how* the indices differ.

Making arguments about performance that don’t include a measurement is premature optimization. But I can assure you: if you put enough matrices into a mapping that 10µs matter, you have bigger performance problems:

```
In [1]: import pandas as pd, numpy as np

In [2]: i2 = (i1 := pd.Index(np.random.rand(10000))).copy()

In [3]: %timeit pd.testing.assert_index_equal(i1, i2)
22.3 µs ± 126 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [4]: %timeit i1.equals(i2)
12.7 µs ± 41.1 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

Reply:
> If you want, I can change it so it calls .equals() first and uses the output from assert_index_equals

Yes.

I wouldn't expect them to put priority on performance on testing functions in the future, and for sure in the past haven't.

Reply:
```suggestion
        if (
            hasattr(val, "index")
            and isinstance(val.index, cabc.Collection)
            and not val.index.equals(self.dim_names)
        ):
            # Could probably also re-order index if it’s contained
            raise ValueError(
                f"value.index does not match parent’s axis {self.axes[0]} names"
            )
```